### DISCUSSION TITLE
ReverseMap with AsProxy

### DISCUSSION THREAD

**Discussion by korser1** [LuckyPennySoftware/AutoMapper#3889] [Ideas]
What do you think about making AsProxy method returning TMappingExpression, and then support ReverseMap?

### Example

```csharp

class Srource {
   public int Value { get; set; }
}

interface IDestination {
   int Value { get; set; }
}


CreateMap<Source, IDestination>().AsProxy().ReverseMap();

**Comment by lbargaoanu** [comment]
I guess it could work. A PR is welcome. Probably I thought it should stand out as the last in the chain.
That would be a binary breaking change, so it implies a major version bump cc @jbogard.

**Reply by korser1** [reply]
I don't see any problems if adding not as the last. Current implementation just adds type map action.
Moreover, we can have the construction like this:


### Source/destination types

```csharp

interface ISource {
   int Value { get; set; }
}

interface IDestination {
   int Value { get; set; }
}


CreateMap<ISource, IDestination>()
     .AsProxy() // proxy for IDestination
     .ReverseMap()
     .AsProxy();  // proxy for ISource

**Reply by lbargaoanu** [reply]
Some things make sense after `AsProxy`, some don't. `ConstructUsing`, `As`, `ConvertUsing` don't make sense for example. I don't want to go through the whole list :) Also having it buried somewhere in the mapping seems wrong (`ReverseMap` only makes things worse). But in the end, whatever :) If @jbogard agrees to bump the major version, you can open a PR.

**Reply by jbogard** [reply]
What does AsProxy return today?

**Reply by korser1** [reply]
> What does AsProxy return today?

void

**Reply by jbogard** [reply]
So would this be a breaking change then?

**Reply by lbargaoanu** [reply]
https://docs.microsoft.com/en-us/dotnet/core/compatibility/
API Compat would warn us.
Code compiled against the old version would throw `MissingMethodException` when ran against the new AM version.

**Reply by korser1** [reply]
> So would this be a breaking change then?

Imagine that someone uses AsProxy now. They cannot chain anything after AsProxy call. And we are not breaking anything for them.

The implementation of the method will not be changed.
But fluent syntax will work for interfaces as for classes.

classes:
CreateMap<Class1, Class2>().ForMember(...).ReverseMap().ForMember(...)

interfaces:
CreateMap<Interface1, Interface2>().ForMember(...).**AsProxy()**.ReverseMap().ForMember(...).**AsProxy()**

Before version 11 we didn't use AsProxy as it was automatically in effect.

**Reply by jbogard** [reply]
It's a binary breaking change in that it's not a drop-in replacement. You'd have to re-compile

**Reply by korser1** [reply]
> 

Sure

**Reply by lbargaoanu** [reply]
@korser1 https://docs.microsoft.com/en-us/dotnet/core/compatibility/categories
You're talking about source compatibility, but we have to maintain binary compatibility for the same the major version.

**Reply by korser1** [reply]
> @korser1 https://docs.microsoft.com/en-us/dotnet/core/compatibility/categories You're talking about source compatibility, but we have to maintain binary compatibility for the same the major version.

Yes, understood. So would you like to have this capability in the next version? I played a little bit, looks good. Will prepare PR.
